File: js-port/funcscript-js/src/funcs/text/lower-text-function.js
const { BaseFunction, CallType } = require('../../core/function-base');
const helpers = require('../helpers');
const { FSDataType } = require('../../core/fstypes');

class LowerTextFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'lower';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 1;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, this.maxParameters);
    if (error) {
      return error;
    }

    const value = helpers.ensureTyped(parameters.getParameter(provider, 0));
    if (helpers.typeOf(value) === FSDataType.Null) {
      return helpers.typedNull();
    }
    if (helpers.typeOf(value) !== FSDataType.String) {
      return helpers.makeError(helpers.FsError.ERROR_TYPE_MISMATCH, `${this.symbol}: string parameter expected`);
    }

    const text = helpers.valueOf(value);
    return helpers.makeValue(FSDataType.String, text.toLowerCase());
  }

  parName(index) {
    return index === 0 ? 'text' : '';
  }
}

module.exports = {
  LowerTextFunction
};

File: FuncScript/Functions/Text/LowerTextFunction.cs
using FuncScript.Core;
using FuncScript.Model;

namespace FuncScript.Functions.Text
{
    [ProviderCollection("text")]
    public class LowerTextFunction : IFsFunction
    {
        public int MaxParsCount => 1;

        public CallType CallType => CallType.Prefix;

        public string Symbol => "lower";

        public int Precedence => 0;

        public object Evaluate(object par)
        {
            var pars = FunctionArgumentHelper.ExpectList(par, this.Symbol);

            
            if (pars.Length != 1)
                return new FsError(FsError.ERROR_PARAMETER_COUNT_MISMATCH, $"{this.Symbol}: single string parameter expected");

            var value = pars[0];

            if (value == null)
                return null;

            if (value is not string text)
                return new FsError(FsError.ERROR_TYPE_MISMATCH, $"{this.Symbol}: string parameter expected");

            return text.ToLowerInvariant();
        }

        public string ParName(int index)
        {
            return index == 0 ? "text" : string.Empty;
        }
    }
}
